engines/adl/adl.cpp

```diff
@@ -0,0 +1,1038 @@
+	switch (f) {
+	case kSupportsLoadingDuringRuntime:
+	case kSupportsSavingDuringRuntime:
+	case kSupportsRTL:
+		return true;
+	default:
+		return false;
+	}
+}
+
+Common::Error AdlEngine::loadGameState(int slot) {
+	Common::String fileName = Common::String::format("%s.s%02d", _targetName.c_str(), slot);
+	Common::InSaveFile *inFile = getSaveFileManager()->openForLoading(fileName);
+
+	if (!inFile) {
+		warning("Failed to open file '%s'", fileName.c_str());
+		return Common::kUnknownError;
+	}
+
+	if (inFile->readUint32BE() != MKTAG('A', 'D', 'L', ':')) {
+		warning("No header found in '%s'", fileName.c_str());
+		delete inFile;
+		return Common::kUnknownError;
+	}
+
+	byte saveVersion = inFile->readByte();
+	if (saveVersion != SAVEGAME_VERSION) {
+		warning("Save game version %i not supported", saveVersion);
+		delete inFile;
+		return Common::kUnknownError;
+	}
+
+	// Skip description
+	inFile->seek(SAVEGAME_NAME_LEN, SEEK_CUR);
+	// Skip save time
+	inFile->seek(6, SEEK_CUR);
+
+	uint32 playTime = inFile->readUint32BE();
+
+	Graphics::skipThumbnail(*inFile);
+
+	initState();
+
+	_state.room = inFile->readByte();
+	_state.moves = inFile->readByte();
+	_state.isDark = inFile->readByte();
+
+	uint32 size = inFile->readUint32BE();
+	if (size != _state.rooms.size())
+		error("Room count mismatch (expected %i; found %i)", _state.rooms.size(), size);
+
+	for (uint i = 0; i < size; ++i) {
+		_state.rooms[i].picture = inFile->readByte();
+		_state.rooms[i].curPicture = inFile->readByte();
+	}
+
+	size = inFile->readUint32BE();
+	if (size != _state.items.size())
+		error("Item count mismatch (expected %i; found %i)", _state.items.size(), size);
+
+	for (uint i = 0; i < size; ++i) {
+		_state.items[i].room = inFile->readByte();
+		_state.items[i].picture = inFile->readByte();
+		_state.items[i].position.x = inFile->readByte();
+		_state.items[i].position.y = inFile->readByte();
+		_state.items[i].state = inFile->readByte();
+	}
+
+	size = inFile->readUint32BE();
+	if (size != _state.vars.size())
+		error("Variable count mismatch (expected %i; found %i)", _state.vars.size(), size);
+
+	for (uint i = 0; i < size; ++i)
+		_state.vars[i] = inFile->readByte();
+
+	if (inFile->err() || inFile->eos())
+		error("Failed to load game '%s'", fileName.c_str());
+
+	delete inFile;
+
+	setTotalPlayTime(playTime);
+
+	_isRestoring = true;
+	return Common::kNoError;
+}
+
+bool AdlEngine::canLoadGameStateCurrently() {
+	return _canRestoreNow;
+}
+
+Common::Error AdlEngine::saveGameState(int slot, const Common::String &desc) {
+	Common::String fileName = Common::String::format("%s.s%02d", _targetName.c_str(), slot);
+	Common::OutSaveFile *outFile = getSaveFileManager()->openForSaving(fileName);
+
+	if (!outFile) {
+		warning("Failed to open file '%s'", fileName.c_str());
+		return Common::kUnknownError;
+	}
+
+	outFile->writeUint32BE(MKTAG('A', 'D', 'L', ':'));
+	outFile->writeByte(SAVEGAME_VERSION);
+
+	char name[SAVEGAME_NAME_LEN] = { };
```

Comment:
minor: I think it would make sense to remove the default initialization and simply call `Common::strlcpy` instead of `strncpy` below.